Kokkos versions of bispectrum and Gaussian descriptors on grid

RandomDefaultUser opened the pull request:
**Summary**

A while back @athomps implemented the calculation of bispectrum descriptors, based on the SNAP formalism, into LAMMPS. This was done as part of the MALA machine learning workflow (results have been published in a few papers already, for a good overview see, e.g., https://arxiv.org/abs/2411.19617). This PR introduces a Kokkos version of the bispectrum descriptor calculation on the grid. It further introduces the calculation of Gaussian descriptors (Gaussian functions centered on atoms, calculated per grid point), for which also a Kokkos version exists. These Gaussian descriptors are needed within MALA to efficiently evaluate the structure factor. With Kokkos, both calculations can be done with multiple GPUs, which is crucial for the MALA workflow .

**Related Issue(s)**

None. Also, this implementation has closed multiple issues in MALA https://github.com/mala-project/mala. 

**Author(s)**

Andrew Rohskopf @rohskopf (at the time of implementation, Sandia National Labs)
Aidan Thompson @athomps (Sandia National Labs)
Lenz Fiedler @RandomDefaultUser (me, currently at Helmholtz-Zentrum Dresden Rossendorf, Germany)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

This PR is backwards compatible.

**Implementation Notes**

Verifying the correctness of results with just LAMMPS. The correctness of the results has been extensively studied with MALA though, the results can be found in https://arxiv.org/abs/2411.19617 and https://nbn-resolving.org/urn:nbn:de:bsz:14-qucosa2-936753. I can provide (closed) PRs from the MALA project as well if that helps. 

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included (should I do that? This is not really a feature regular users will use)
- [x] The added/updated documentation is integrated and tested with the documentation build system (should I do that? This is not really a feature regular users will use)
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree. (I am not sure if we should add something here?)
- [x] A package specific README file has been included or updated (should I do that? This is not really a feature regular users will use)
- [x] One or more example input decks are included

RandomDefaultUser committed:
Re-starting MALA branch in MALA fork

rohskopf committed:
Start ComputeSNAGridKokkos implementation

rohskopf committed:
Experiment with different implementations

rohskopf committed:
Remove swo

rohskopf committed:
Mimic pair snap kokkos pattern

rohskopf committed:
Remove swo

rohskopf committed:
Add all SNAP computations and 4D view Kokkos memory allocator

rohskopf committed:
Sync device and host compute arrays

rohskopf committed:
Debug inability to deep copy

rohskopf committed:
Change ntotal to n_ninside

rohskopf committed:
Add cutoff view properly

rohskopf committed:
Matching descriptors when no neighbors, good checkpoint for debugging

rohskopf committed:
Fix neighbor criteria so atoms sharing positions with gridpoints aren't included as neighbors

rohskopf committed:
Agreement between Kokkos and original compute sna/grid with switchflag = 1; note that switchflag = 0 gives wrongly zeroed values for Kokkos because of bug in compute_s_dsfac function of sna_kokkos_impl.h causing sfac to be zero

rohskopf committed:
Organize sna method calls and clean up

rohskopf committed:
Organize deallocator calls

RandomDefaultUser committed:
Merge branch 'lammps:develop' into mala

rohskopf committed:
Replace limited parallel for with normal for for now

rohskopf committed:
Use normal loop over ntotal inside neighbor team policy

rohskopf committed:
Clean up kernels

rohskopf committed:
Loop over chunks on GPU to write values properly when using default chunk size

rohskopf committed:
Fix cutoff factor when switchflag = 0

rohskopf committed:
Add triclinic cell conversion

RandomDefaultUser committed:
Re-Adding Gaussian grid again, originally authored by Aidan Thompson

Co-authored-by: Aidan Thompson <athomps@sandia.gov>

RandomDefaultUser committed:
Merge remote-tracking branch 'fork_drew/sna-grid-kokkos' into sna-grid-kokkos

RandomDefaultUser committed:
Merge pull request #1 from rohskopf/sna-grid-kokkos

compute sna/grid/kokkos

rohskopf committed:
Initial Gaussian grid implementation

rohskopf committed:
Prevent polymorphic destructor calls with copymode

rohskopf committed:
Shallow copy Kokkos written array to returned array variable

rohskopf committed:
Fill grid geometry info

rohskopf committed:
Tweak TeamPolicy settings for speedup

rohskopf committed:
Clean up debug prints

RandomDefaultUser committed:
Merge pull request #2 from rohskopf/sna-grid-kokkos

Kokkos Gaussian grid

rohskopf committed:
Initial sna/grid/local/kk implement

rohskopf committed:
Destruct sna/grid/local/kk properly and use local arrays

rohskopf committed:
Fill local sna/grid array

RandomDefaultUser committed:
Merge branch 'lammps:develop' into mala

rohskopf committed:
Remove destructor prints

RandomDefaultUser committed:
Hotfixing a small bug in the Kokkos Gaussian Compute

Co-authored-by: Drew Rohskopf <<drew.rohskopf@gmail.com>

RandomDefaultUser committed:
Merge pull request #3 from rohskopf/sna-grid-kokkos

Kokkos SNAP grid local compute

RandomDefaultUser committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into mala

# Conflicts:
#	src/KOKKOS/kokkos_type.h

RandomDefaultUser committed:
Renamed files to be more consistent with other examples

RandomDefaultUser commented:
I see that there are some tests failing, I was not aware that there would be issues. Good thing I put the PR as draft, I will fix that before marking it ready for review.

RandomDefaultUser committed:
Fixed formatting issues

athomps commented:
Formatting issues are fixed. All checks have passed. So I have switched this to ready for review.

athomps reviewed (commented):
This looks very solid. However, the doc pages will need to be updated.  In partcular, a new doc page for the Gaussian descriptor needs to be created at doc/src/compute_gaussian_grid(_local).rst. I suggest stripping down and editing doc/src/compute_sna_atom.rst.

RandomDefaultUser committed:
Added documentation

RandomDefaultUser commented:
Ah you are right @athomps, forgot about the docs. I added them now, let me know if there is anything missing.

RandomDefaultUser committed:
Added link to PRB paper

RandomDefaultUser committed:
Added logs for examples

RandomDefaultUser committed:
Now actually added the correct log

RandomDefaultUser committed:
Docs and example updates by Aidan Thompson

I did NOT author this commit, I am only pushing it because for some reason, Github does not permit Aidan to do so

Co-authored-by: Aidan Thompson <athomps@sandia.gov>

RandomDefaultUser commented:
@athomps I added your commit as discussed!

athomps reviewed (commented):
I approve

RandomDefaultUser committed:
Merge branch 'develop' into mala

RandomDefaultUser committed:
Fixed two style issues in the docs, got rid of printf that's also deleted on develop

RandomDefaultUser commented:
There were two minor style issues in the commit updating the docs, I fixed them, we should be got to go now.

RandomDefaultUser committed:
I forgot to include a change in merging develop that seems to be very important

RandomDefaultUser committed:
Code by Aidan Thompson, I am only committing it; fixing the cyclical include that broke the build process after merging develop

Co-authored-by: Aidan Thompson <athomps@sandia.gov>

akohlmey commented:
There is a problem compiling with some platforms.

stanmoore1 commented:
>There is a problem compiling with some platforms.

@akohlmey yes, this came from merging https://github.com/lammps/lammps/pull/4391. @weinbe2 and I will work on this after the break.

akohlmey commented:
> > There is a problem compiling with some platforms.
> 
> @akohlmey yes, this came from merging #4391. @weinbe2 and I will work on this after the break.

@stanmoore1 Just a gentle reminder that this is still pending.

stanmoore1 committed:
Add new files to GNU Make build system

stanmoore1 committed:
Fix some compile issues and remove unused variables

stanmoore1 committed:
Fix compile issues from #4391

stanmoore1 committed:
Fix issues with host_flag

stanmoore1 committed:
Merge branch 'develop' of github.com:lammps/lammps into mala

stanmoore1 commented:
@RandomDefaultUser thank you for this contribution. I fixed the compile errors from merging https://github.com/lammps/lammps/pull/4391 in my branch here: https://github.com/stanmoore1/lammps/tree/mala. I don't have permission to push to your branch, and couldn't seem to create a PR either. Can you please merge my changes into your branch?

@weinbe2 FYI

stanmoore1 commented:
@RandomDefaultUser actually Axel showed me how to submit a PR to your fork, see https://github.com/mala-project/lammps/pull/6.

RandomDefaultUser commented:
Hi @stanmoore1 thank you so much for fixing the compile issues! I will merge the PR you created right away so that we can then hopefully merge this one as well!

RandomDefaultUser committed:
Merge pull request #6 from stanmoore1/mala

Fix compile issues from LAMMPS https://github.com/lammps/lammps/pull/4391

RandomDefaultUser committed:
Merge branch 'develop' into mala

athomps reviewed (approved):
I approve.

stanmoore1 commented on src/KOKKOS/compute_sna_grid_local_kokkos.cpp:
```diff
@@ -0,0 +1,81 @@
+
+
+
+// The following chunk will compile but we're gonna try a wrapper approach like pair snap.
```
For example: please remove extra comments like this one.

stanmoore1 commented on src/KOKKOS/compute_sna_grid_kokkos_impl.h:
```diff
@@ -0,0 +1,916 @@
+  // The purpose here is to transform for triclinic boxes.
+  /*
+  if (triclinic){
+    printf("We are triclinic %f %f %f\n", xtmp, ytmp, ztmp);
```
For example: remove debug comments here.

stanmoore1 reviewed (changes requested):
@RandomDefaultUser Thank you again for this contribution. Overall the code looks good. My only request is to remove the unneeded commented out parts of the code so it is a little cleaner and easier to read. A few comments are fine but please remove large blocks of unused code. Thanks!

RandomDefaultUser committed:
Removed debug comments/old printfs etc.

RandomDefaultUser commented:
Hi @stanmoore1 , thanks for the feedback, you're right, there were some leftovers from debugging in the code. I removed them, let me know if there are still comments you think are unnecessary and should maybe be removed.

RandomDefaultUser committed:
Merge branch 'develop' into mala

athomps commented:
This is ready to go, I think! @stanmoore1 do you have any concerns?

stanmoore1 commented on src/KOKKOS/Install.sh:
```diff
@@ -119,6 +119,12 @@ action compute_composition_atom_kokkos.cpp compute_composition_atom.cpp
+action compute_sna_grid_kokkos_impl.h compute_sna_grid.cpp
+action compute_sna_grid_local_kokkos.cpp compute_sna_grid_local.cpp
+action compute_sna_grid_local_kokkos.h compute_sna_grid_local.h
+action compute_sna_grid_local_kokkos_impl.h compute_sna_grid_local.cpp
```
Need to add the `compute_gaussian_grid_local_kokkos.*` files too here.

RandomDefaultUser replied:
Got it, I have added the files!

stanmoore1 commented on src/KOKKOS/compute_sna_grid_kokkos_impl.h:
```diff
@@ -0,0 +1,786 @@
+------------------------------------------------------------------------- */
+
+/* ----------------------------------------------------------------------
+   Contributing authors: Christian Trott (SNL), Stan Moore (SNL),
```
@RandomDefaultUser Please feel free to remove our names and add yours since you created this file.

RandomDefaultUser replied:
I credited @rohskopf because he created the file, I am only taking care of the PR. But thanks for pointing this out!

stanmoore1 commented on src/KOKKOS/pair_mliap_kokkos.cpp:
```diff
@@ -233,13 +233,15 @@ void PairMLIAPKokkos<DeviceType>::coeff(int narg, char **arg) {
   // map[i] = which element the Ith atom type is, -1 if not mapped
   // map[0] is not used
 
+  //printf(">>> ntypes: %d\n", atom->ntypes);
```
I think this is debugging output, please remove.

RandomDefaultUser replied:
Got it, I removed it!

stanmoore1 reviewed (changes requested):
@RandomDefaultUser I missed a file in `/src/KOKKOS/Install.sh`, can you please add that? I had a few other minor change requests too, otherwise this is ready to merge.

RandomDefaultUser committed:
Removed debugging output, added contributor, added files to Install.sh

RandomDefaultUser committed:
Merge branch 'develop' into mala

RandomDefaultUser commented:
I addressed the small change requests you had @stanmoore1, let me know if there is anything else!

stanmoore1 commented:
Doing a quick test on GPU now.

stanmoore1 committed:
Fix issues with GNU Make build

stanmoore1 commented:
GPU tests look fine, but I found a couple more issues with the GNU Make build. @RandomDefaultUser can you merge my PR here: https://github.com/mala-project/lammps/pull/7? Thanks.

RandomDefaultUser committed:
Merge pull request #7 from stanmoore1/mala

Fix issues with GNU Make build

RandomDefaultUser commented:
Of course, I just merged the PR!

RandomDefaultUser committed:
Merge branch 'develop' into mala

stanmoore1 commented:
This PR is ready to merge from my POV.
